### PR TITLE
CFN Publish Integration

### PR DESCRIPTION
**Pull request by HieronymusLex**
Staging bucket which acts as the integration point for CFN Publish https://github.com/aws-solution-builders/cfn-publish

Workflow:
1. Travis executes test etc. as normal then runs `git archive -o <project>.zip`
2. Travis uploads <project>.zip to the staging bucket under `<project>/<version>/<project>.zip`
3. Lambda invokes `cfn-publish` build job with source override and version in environment vars
4. CFN publish takes care of replication

### CODE REVIEW

File: templates/staging_bucket.template [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+            Bucket: !Ref SrcBucket
+            Events: s3:ObjectCreated:*
+            Filter:
+              S3Key:
+                Rules:
+                  - Name: suffix
+                    Value: .zip
+
+  SrcBucket:
+    Type: AWS::S3::Bucket
```
Comment (HieronymusLex) [inline templates/staging_bucket.template]: perhaps it makes sense to have some lifecycle rules here since it's effectively a staging bucket?

**Comment by HieronymusLex** [comment]
is uploading to `<project>/<version>/<project>.zip` a bit weird?? Options are somewhat limited since in cfn-publish we can only really get the project name from the last segment of the object key (ref: https://github.com/aws-solution-builders/cfn-publish/pull/2/files)

File: templates/staging_bucket.template (hunk 1 above)
Comment (svozza) [reply templates/staging_bucket.template]: Yeah could let them disappear after a few weeks maybe.

File: README.md [hunk 2]
```diff
@@ -1,8 +1,62 @@
...
+
+## Deployment
+
+1. Clone this repo
+2. Set the relevant vars in your shell:
+
+```bash
+TEMP_BUCKET=some-temp-cfn-bucket # Temp bucket to upload this stacks artefacts to
+STACK_NAME=staging-bucket # Name you want for this stack
+CFN_PUBLISH_PROJECT=cfn-publish # The name of your cfn-publish CodeBuild job
```
Comment (stilvoid) [inline README.md]: `s/job/project/`

File: README.md [hunk 3]
```diff
@@ -1,8 +1,62 @@
 ## Solutions Build Tools
 
-Tools for releasing solutions in multiple regions in an opinionated way
+Tools for releasing solutions in multiple regions via [cfn-publish]
```
Comment (stilvoid) [inline README.md]: You know what, I had no idea you can do links this way :D

File: README.md [hunk 4]
```diff
@@ -1,8 +1,62 @@
...
+```bash
+BUCKET=$(aws cloudformation describe-stacks --stack-name $STACK_NAME --query 'Stacks[0].Outputs[?OutputKey==`Bucket`].OutputValue' --output text)
+aws s3 cp ./$VERSION.zip s3://$BUCKET/$PROJECT_NAME/$VERSION/$PROJECT_NAME.zip
+```
+
+You can now check the cfn-publish project in CodeBuild to verify cfn-publish has executed successfully.
+
+## Failure Notifications
+An SNS topic is provisioned to act as the DLQ for the Lambda trigger which you can subscribe to. If there is a problem processing an upload to the staging bucket,
+a notification will be published to this topic.
```
Comment (stilvoid) [inline README.md]: Really need to set an SNS topic up for cfn-publish too

File: lambdas/index.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        raise ValueError("CODEBUILD_PROJECT environment variable not set")
+
+    for record in event["Records"]:
+        try:    
+            bucket = record["s3"]["bucket"]["name"]
+            object_key = record["s3"]["object"]["key"]
+            build_data = get_build_data(object_key)
+            start_build(build_data["version"], "{bucket}/{key}".format(bucket=bucket, key=build_data["object_key"]))
+        except IndexError as e:
+            logger.error("Unable to determine artefact project and version from object key: {}. Uploads should be in the format s3://BUCKET/PROJECT/VERSION/PROJECT.zip".format(object_key))
```
Comment (stilvoid) [inline lambdas/index.py]: Why the redundant `PROJECT`? Could just be `s3://BUCKET/PROJECT/VERSION.zip`?

File: templates/staging_bucket.template [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        - Version: '2012-10-17'
+          Statement:
+          - Effect: Allow
+            Action:
+              - "codebuild:StartBuild"
+              - "s3:Get*"
+              - "s3:List*"
+            Resource: "*"
+      Events:
+        ArtifactUploaded:
```
Comment (stilvoid) [inline templates/staging_bucket.template]: You've spelled it "Art**e**fact" elsewhere.

File: lambdas/index.py (hunk 5 above)
Comment (HieronymusLex) [reply lambdas/index.py]: @stilvoid it's a limitation of `cfn-publish` unfortunately. currently cfn-publish (unless you provide your own value in a config file) gets the project name by taking the last segment of the CODEBUILD_SOURCE_REPO_URL and stripping the extension e.g.:
- git@github.com:aws-solution-builders/solutions-build-tools.git -> solutions-build-tools
- s3://bucket/my/object/key.zip -> key

I'd be up for changing it but i don't see any other sensible way for `cfn-publish` to derive the project name by default?

File: README.md (hunk 4 above)
Comment (HieronymusLex) [reply README.md]: yeah - though we also need to find a way for it to fail the jobs when something goes wrong too!

File: templates/staging_bucket.template (hunk 6 above)
Comment (HieronymusLex) [reply templates/staging_bucket.template]: Updated

File: templates/staging_bucket.template (hunk 1 above)
Comment (HieronymusLex) [reply templates/staging_bucket.template]: I've updated it to set the expiry to 14 days

**Pull request merged**